File: Tests/CorePromise/bridging.test.m
#import <PromiseKit/PromiseKit.h>
@import XCTest;

__attribute__((objc_runtime_name("PMKPromiseBridgeHelper")))
__attribute__((objc_subclassing_restricted))
@interface PromiseBridgeHelper: NSObject
- (AnyPromise *)bridge1;
@end


@interface BridgingTestCase_ObjC: XCTestCase @end @implementation BridgingTestCase_ObjC

- (void)test1 {
    XCTestExpectation *ex = [self expectationWithDescription:@""];
    AnyPromise *promise = PMKAfter(0.02);
    for (int x = 0; x < 100; ++x) {
        promise = promise.then(^{
            return [[[PromiseBridgeHelper alloc] init] bridge1];
        });
    }
    promise.then(^{
        [ex fulfill];
    });
    [self waitForExpectationsWithTimeout:10 handler:nil];
}

@end


@implementation PromiseBridgeHelper (objc)

- (AnyPromise *)bridge2 {
    return [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            resolve(@123);
        });
    }];
}

@end

File: Tests/CorePromise/bridging.test.swift
import Foundation
import PromiseKit
import XCTest

class BridgingTestCase_Swift: XCTestCase {
    func testCanBridgeAnyObject() {
        let sentinel = NSURLRequest()
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)

        XCTAssertEqual(ap.value(forKey: "value") as? URLRequest, sentinel)
    }

    func testCanBridgeOptional() {
        let sentinel: NSURLRequest? = NSURLRequest()
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)

        XCTAssertEqual(ap.value(forKey: "value") as? URLRequest, sentinel!)
    }

    func testCanBridgeSwiftArray() {
        let sentinel = [NSString(),NSString(),NSString()]
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)

        XCTAssertEqual(ap.value(forKey: "value") as! [NSString], sentinel)
    }

    func testCanBridgeSwiftDictionary() {
        let sentinel = [NSString():NSString()]
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)

        XCTAssertEqual(ap.value(forKey: "value") as! [NSString:NSString], sentinel)
    }

    func testCanBridgeInt() {
        let sentinel = 3
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)
        XCTAssertEqual(ap.value(forKey: "value") as? Int, sentinel)
    }

    func testCanBridgeString() {
        let sentinel = "a"
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)
        XCTAssertEqual(ap.value(forKey: "value") as? String, sentinel)
    }

    func testCanBridgeBool() {
        let sentinel = true
        let p = Promise(sentinel)
        let ap = AnyPromise(bound: p)
        XCTAssertEqual(ap.value(forKey: "value") as? Bool, sentinel)
    }

    func testCanChainOffAnyPromiseReturn() {
        let ex = expectation(withDescription: "")

        firstly {
            Promise(1)
        }.then { _ -> AnyPromise in
            return PromiseBridgeHelper().value(forKey: "bridge2") as! AnyPromise
        }.then { value -> Void in
            XCTAssertEqual(123, value as? Int)
            ex.fulfill()
        }

        waitForExpectations(withTimeout: 1, handler: nil)
    }


    func testCanThenOffAnyPromise() {
        let ex = expectation(withDescription: "")

        let ap = PMKDummyAnyPromise_YES() as! AnyPromise
        ap.then { obj -> Void in
            if let value = obj as? NSNumber {
                XCTAssertEqual(value, NSNumber(value: true))
                ex.fulfill()
            }
        }

        waitForExpectations(withTimeout: 1, handler: nil)
    }

    func testCanThenOffManifoldAnyPromise() {
        let ex = expectation(withDescription: "")

        let ap = PMKDummyAnyPromise_Manifold() as! AnyPromise
        ap.then { obj -> Void in
            if let value = obj as? NSNumber {
                XCTAssertEqual(value, NSNumber(value: true))
                ex.fulfill()
            }
        }

        waitForExpectations(withTimeout: 1, handler: nil)
    }
}


// for bridging.test.m
@objc(PMKPromiseBridgeHelper) class PromiseBridgeHelper: NSObject {
    @objc func bridge1() -> AnyPromise {
        let p = after(0.01)
        return AnyPromise(bound: p)
    }
}
